### PR TITLE
Fix performance regression in job submission.

### PR DESCRIPTION
**Pull request by bdice**
## Description
Corrects a serious performance regression in job submission, introduced in 7e2722fe25bf5fcf3009a5b77e28e187a5eb8050.
Makes sure the `operations` generator is consumed *inside* the buffering context manager.

I also added a second buffered region during submission, so that status updates do not immediately write to the project document (only when finished, an error occurs, or the process is gracefully interrupted, all of which will trigger the context manager's exit and cause a flush to disk). Without buffering, this region is very expensive in large projects (which need to read/write the project document for every job that is submitted).

## Motivation and Context
Submission was very slow in a large workspace (100k jobs).

## Types of Changes

- [ ] Documentation update
- [x] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac-flow/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac-flow/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md#code-style) of this project.
- [ ] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [x] I have updated the [changelog](https://github.com/glotzerlab/signac-flow/blob/master/changelog.txt).

### CODE REVIEW

**Review by b-butler — APPROVED** [review]
@bdice thanks for the fix for the regression I introduced.

**Comment by bdice** [comment]
@b-butler I added a second region of buffering, which should still be safe-ish and should provide significant performance benefits. However, I'm going to re-request your review since this change is "new" code and not just a fix. I will verify the performance separately and comment again on this PR to confirm that it's as fast as expected.

File: changelog.txt [hunk 1]
```diff
@@ -1,6 +1,6 @@
-=======
-Changes
-=======
+=========
+Changelog
+=========
```
Comment (csadorf) [inline changelog.txt:3]: I'll allow it. ;)

File: flow/project.py [hunk 2]
```diff
@@ -3027,19 +3027,28 @@ def submit(self, bundle_size=1, jobs=None, names=None, num=None, parallel=False,
         # Gather all pending operations.
         with self._potentially_buffered():
             default_directives = self._get_default_directives()
-            operations = self._get_submission_operations(jobs, default_directives, names,
-                                                         ignore_conditions,
-                                                         ignore_conditions_on_execution)
-        if num is not None:
-            operations = list(islice(operations, num))
+            # The generator must be used *inside* the buffering context manager
+            # for performance reasons
```
Comment (csadorf) [inline flow/project.py:3031]:
```suggestion
            # for performance reasons.
```

**Review by csadorf — APPROVED** [review]
Great, thx! 😃

File: changelog.txt (hunk 1 above)
Comment (bdice) [reply changelog.txt:3]: @csadorf This makes the name consistent with signac. 😉

File: flow/project.py [hunk 3]
```diff
@@ -3027,19 +3027,28 @@ def submit(self, bundle_size=1, jobs=None, names=None, num=None, parallel=False,
...
-            operations = list(islice(operations, num))
+            # The generator must be used *inside* the buffering context manager
+            # for performance reasons.
+            operation_generator = self._get_submission_operations(jobs,
+                                                                  default_directives,
+                                                                  names,
+                                                                  ignore_conditions,
+                                                                  ignore_conditions_on_execution)
+            # islice takes the first "num" elements from the generator, or all
+            # items if num is None
```
Comment (bdice) [inline flow/project.py:3038]:
```suggestion
            # items if num is None.
```

**Comment by bdice** [comment]
I can confirm that this improves performance (I did not record any values, but the output scrolls by faster). There is still significant room for performance improvements in the code that is called before this part. For a submission of ~25,000 operations, signac-flow took nearly 45 minutes to do ... something (no idea what) ... before submission began. I was not able to profile this on the system where this submission took place, and will need to build a large workspace and local testing SLURM server to profile and test on.

**Pull request merged**